Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp

```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
 
     double final;
-    if (relative_end < 0)
+    if (Value(relative_end).is_negative_infinity()) {
+        // 11. If relativeEnd is -∞, let final be 0.
+        final = 0;
+    } else if (relative_end < 0) {
+        // 12. Else if relativeEnd < 0, let final be max(len + relativeEnd, 0).
         final = max(length + relative_end, 0.0);
-    else
+    } else {
+        // 13. Else, let final be min(relativeEnd, len).
         final = min(relative_end, (double)length);
+    }
 
+    // 14. Let newLen be max(final - first, 0).
     auto new_length = max(final - first, 0.0);
 
+    // 15. Let ctor be ? SpeciesConstructor(O, %ArrayBuffer%).
     auto* constructor = TRY(species_constructor(global_object, *array_buffer_object, *global_object.array_buffer_constructor()));
+
+    // 16. Let new be ? Construct(ctor, « 𝔽(newLen) »).
     auto* new_array_buffer = TRY(construct(global_object, *constructor, Value(new_length)));
 
+    // 17. Perform ? RequireInternalSlot(new, [[ArrayBufferData]]).
+    // fixme: is this the exception we want here?
     if (!is<ArrayBuffer>(new_array_buffer))
         return vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorDidNotCreate, "an ArrayBuffer");
     auto* new_array_buffer_object = static_cast<ArrayBuffer*>(new_array_buffer);
 
-    // FIXME: Check for shared buffer
+    // 18. If IsSharedArrayBuffer(new) is true, throw a TypeError exception.
+    // fixme: check for shared buffer
+
+    // 19. If IsDetachedBuffer(new) is true, throw a TypeError exception.
     if (new_array_buffer_object->is_detached())
         return vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorReturned, "a detached ArrayBuffer");
+
+    // 20. If SameValue(new, O) is true, throw a TypeError exception.
     if (same_value(new_array_buffer_object, array_buffer_object))
         return vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorReturned, "same ArrayBuffer instance");
+
+    // 21. If new.[[ArrayBufferByteLength]] < newLen, throw a TypeError exception.
     if (new_array_buffer_object->byte_length() < new_length)
         return vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorReturned, "an ArrayBuffer smaller than requested");
 
+    // 22. NOTE: Side-effects of the above steps may have detached O.
+    // 23. If IsDetachedBuffer(O) is true, throw a TypeError exception.
     if (array_buffer_object->is_detached())
         return vm.throw_completion<TypeError>(global_object, ErrorType::DetachedArrayBuffer);
 
-    // This is ugly, is there a better way to do this?
-    array_buffer_object->buffer().span().slice(first, new_length).copy_to(new_array_buffer_object->buffer().span());
+    // 24. Let fromBuf be O.[[ArrayBufferData]].
+    auto from_buf = array_buffer_object->buffer();
+
+    // 25. Let toBuf be new.[[ArrayBufferData]].
+    auto to_buf = new_array_buffer_object->buffer();
```

Comment:
Same here, don't think this actually works. If we copy the buffer, the returned object is unaffected.

Reply:
Silly me, this is one of those last minute stylistic changes I made - but never ran test-js for.